Acravn/external secrets db support

acravn opened the pull request:
### Description
@Maddosaurus - Tagging you based on the slack thread on the CNCF stackrox slack channel.

 I added some extra functionality to the helm charts in my forked version of the helm chart repo to meet several requirements my organization needed in order to deploy stackrox. 

- Ability to use an external DB with the scanner service such as RDS,etc
- Ability to create an arbitrarily named k8s secret that contains a DB password that can set the DB password for the postgres deployments and pass the password to the central / scanner services
- Ability to create k8s generic secrets ahead of time containing TLS values such as cert/key for the CA that can be passed into the helm chart instead of pasting the values directly in the values file

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

<!--
General Availability requirements: https://github.com/stackrox/stackrox/blob/master/PR_GA.md
Feature Flags usage: https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md
-->

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

<!--
If no tests have been contributed, please explain why unless it's obvious,
e.g., the PR is a one-line comment change.
-->

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

#### Testing External DB for both Central and Scanner DB and pass in externalSecret for db passwords / ca cert&key

1. Create a postgres deployment for central-db and scanner-db
2. Populate scanner-db postgres with `stackrox-io/scanner-db:4.5.0` image's init sql script
3. Generate a CA cert and key with openssl and write to a k8s secret in the stackrox namespace
```
openssl genrsa 2048 > certs/ca-key.pem

openssl req -new -x509 -nodes \
  -key "certs/ca-key.pem" \
  -out "certs/ca.pem" \
  -days 3652 \
  -subj "/CN=StackRox Certificate Authority" \
  -addext "keyUsage=critical,digitalSignature,keyEncipherment,keyCertSign" \
  -addext "extendedKeyUsage=serverAuth,clientAuth"
```

4. Configure a `test-values.yaml` file with the below settings:
```
ca:
  existingSecret: 
    name: central-ca-tls
    cert: ca.pem
    key: ca-key.pem
imagePullSecrets:
  username:  ###UPDATE_ME###
  password: ###UPDATE_ME###
network:
  enableNetworkPolicies: false
env:
  istio: true
central:
  persistence:
    none: true
  db:
    external: true
    existingSecret:
      name: central-db-password
      key: password
    source:
      connectionString: host=central-db port=5432 dbname=central_active user=central sslmode=disable

scanner:
  db:
    external: true
    connectionString: host=scanner-db port=5432 dbname=scanner user=postgres sslmode=disable
    existingSecret: 
      name: scanner-db-password
      key: password
```
5. Install central-services helm chart (`helm upgrade -i stackrox-central-services . -f test-values.yaml`
6. Extract the admin password and pull the init bundle
7. Create a set of secrets to pass in the TLS certs for secured-cluster-services
8. Configure the below `test-values.yaml` file
```
clusterName: lighthouse
centralEndpoint: central.stackrox.svc:443
imagePullSecrets:
  username: ###UPDATE_ME###
  password: ###UPDATE_ME###

env:
  istio: true

ca:
  existingSecret: 
    name: central-ca-tls
    cert: ca.pem

sensor:
  serviceTLS:
    existingSecret: 
      name: lh-sensor-tls
      cert: cert.pem
      key: key.pem

admissionControl:
  serviceTLS:
    existingSecret: 
      name: lh-admission-control-tls
      cert: cert.pem
      key: key.pem

collector:
  serviceTLS:
    existingSecret:
      name: lh-collector-tls
      cert: cert.pem
      key: key.pem

```
9. Deploy the helm chart - `helm upgrade -i stackrox-secured-cluster-services . -f test-values.yaml`

#### Testing Central and Scanner DB with provided DB password
1. Use the below `test-values.yaml` to install central services
```
imagePullSecrets:
  username:  ###UPDATE_ME###
  password: ###UPDATE_ME###
network:
  enableNetworkPolicies: false
env:
  istio: true
central:
  persistence:
    none: true
  db:
    persistence:
      persistentVolumeClaim: 
        size: '15Gi'
    resources:
     requests:
       memory: "4Gi"
       cpu: "1"
     limits:
       memory: "4Gi"
       cpu: "1"
    # external: true
    existingSecret:
      name: central-db-password
      key: password
    # source:
    #   connectionString: host=central-db port=5432 dbname=central_active user=central sslmode=disable

scanner:
  db:
    # external: true
    # connectionString: host=scanner-db port=5432 dbname=scanner user=postgres sslmode=disable
    existingSecret: 
      name: scanner-db-password
      key: password
```

#### Testing Central and Scanner deployment without modifications
1. Use the below values file:
```
imagePullSecrets:
  username:  ###UPDATE_ME###
  password: ###UPDATE_ME###
network:
  enableNetworkPolicies: false
env:
  istio: true
central:
  persistence:
    none: true
  db:
    persistence:
      persistentVolumeClaim: 
        size: '15Gi'
    resources:
     requests:
       memory: "4Gi"
       cpu: "1"
     limits:
       memory: "4Gi"
       cpu: "1"
```

Maddosaurus commented:
/ok-to-test

Maddosaurus commented:
Hey Andrew,

thanks for the PR!
I've tagged some additional folks from Engineering in, which can review the changes to their respective components.

@stackrox/scanner For the changes to scanner DB
@stackrox/core-workflows For the changes to central DB
And last but not least, @mclasmeier for the changes to the Helm charts and secret management.

From what I can see, the changes look good! Additional unit tests for the new behavior would be a welcome addition, though.
I would recommend to hold with these until the teams have discussed whether they want the new functionality included in the charts, though, to be efficient with everyones' time :smile:

acravn commented:
Thanks @Maddosaurus, I'll be happy to write unit tests if the team wants to add the changes.

dashrews78 commented:
From the perspective of central-db, I like this concept.  I think it is worthy to pursue.  I will defer to the helm experts for a more thorough review and scanner to see if they agree in regards to an external version of their database.

flickerfly commented:
@dashrews78 How do we get the right "helm experts" eyes to accept this conceptually before putting more development effort in?

dashrews78 commented:
> @mclasmeier

That would be @mclasmeier whom was already tagged on this.

mclasmeier commented:
I have seen this, will TAL.

mclasmeier commented on image/templates/helm/shared/templates/_crypto.tpl:
```diff
@@ -245,3 +245,27 @@
+    {{ $val := (index $secretVal.data $secretKey) | b64dec }}
+    {{ $_ := set $ $key $val }}
+  {{ else }}
+    {{ $_ := set $ $key "fake_value" }}
```
Seems like a potential cause for some runtime breakage?

acravn replied:
Good point, I was having issues with the helm template command handling the lookup, but I see you made a comment about a safe lookup wrapper function. I'll check that out and clean this up.

mclasmeier commented on image/templates/helm/shared/templates/_crypto.tpl:
```diff
@@ -245,3 +245,27 @@
+{{ $key := index . 2 }}
+{{ with $.existingSecret }}
+  {{ $secret := $.existingSecret.name }}
+  {{ $secretVal := (lookup "v1" "Secret" $ns $secret) }}
```
We have a safe lookup wrapper function.

mclasmeier commented on image/templates/helm/stackrox-secured-cluster/templates/_init.tpl.htpl:
```diff
@@ -36,6 +36,23 @@
+  TLS secrets in a values file that could get committed into source control.
+    */}}
+
+{{ include "srox.getExistingTLSSecret" (list $rox.ca $.Release.Namespace "cert" )}}
```
I am not sure about this approach. I mean, this is essentially doing a `cluster->local->cluster` roundtrip, or am I missing something here? Have you considered to instead implement this in a way that lets you *reference* existing secrets without fetching them and then attempting to recreate them when the rendered manifests get applied to the server?

Also, a caveat: when the Helm chart is rendered within our operator Helm's lookup function is deliberately disabled. This would also suggest to use secret references instead (and skip rendering of the client secrets).

acravn replied:
That's a good point, I was caught up trying to replicate the combination of the CA cert with the service TLS key/cert into a single secret as opposed to updating the templates to reference the existing secret. I'll try to get some time on my end to look into this change in the approach. Some of it is motivated by how we deploy software in airgapped environments and being able to manage secrets separately from the helm chart.

mclasmeier reviewed (commented):
Hi,

thank you very much for your contribution! We appreciate it.
This is just a first round of comments.

But there is on critical open questions that I haven't addressed as part of my comments:

As you have probably noticed there is a new scanner, named scanner V4, which will replace scanner v2 entirely in the not-so-distant future. I have reached out to the scanner regarding this, but I would assume that it would be a much more sustainable feature addition to the scanner V4 templating. We will follow-up on this.
Would you generally be interested in doing the changes to scanner V4 as well / instead?

The other conceptual question we should address is the fetching of pre-existing secrets (I have raised this in the comments).

Again, thank you very much.
Moritz

RTann commented:
> As you have probably noticed there is a new scanner, named scanner V4, which will replace scanner v2 entirely in the not-so-distant future. I have reached out to the scanner regarding this, but I would assume that it would be a much more sustainable feature addition to the scanner V4 templating. We will follow-up on this.
Would you generally be interested in doing the changes to scanner V4 as well / instead?

As @mclasmeier we are in the process of deprecating the older StackRox Scanner (Scanner v2) in favor of the newer Scanner V4. It's definitely preferred to do this for Scanner V4 instead of the older Scanner

flickerfly commented:
> > As you have probably noticed there is a new scanner, named scanner V4, which will replace scanner v2 entirely in the not-so-distant future. I have reached out to the scanner regarding this, but I would assume that it would be a much more sustainable feature addition to the scanner V4 templating. We will follow-up on this.
> > Would you generally be interested in doing the changes to scanner V4 as well / instead?
> 
> As @mclasmeier we are in the process of deprecating the older StackRox Scanner (Scanner v2) in favor of the newer Scanner V4. It's definitely preferred to do this for Scanner V4 instead of the older Scanner

I'm not sure I understand where code is for scanner v2 versus scanner v4 after looking around a bit. Is this documented somewhere or do you have some pointers to get me going the right direction?

acravn commented on image/templates/helm/shared/config-templates/scanner/config.yaml.tpl:
```diff
@@ -16,6 +16,9 @@ scanner:
     options:
       # PostgreSQL Connection string
       # https://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-CONNSTRING
+      {{ if ._rox.scanner.db.external }}
```
I'll update with the left hand side whitespace

RTann commented:
> > > As you have probably noticed there is a new scanner, named scanner V4, which will replace scanner v2 entirely in the not-so-distant future. I have reached out to the scanner regarding this, but I would assume that it would be a much more sustainable feature addition to the scanner V4 templating. We will follow-up on this.
> > > Would you generally be interested in doing the changes to scanner V4 as well / instead?
> > 
> > 
> > As @mclasmeier we are in the process of deprecating the older StackRox Scanner (Scanner v2) in favor of the newer Scanner V4. It's definitely preferred to do this for Scanner V4 instead of the older Scanner
> 
> I'm not sure I understand where code is for scanner v2 versus scanner v4 after looking around a bit. Is this documented somewhere or do you have some pointers to get me going the right direction?

If you are curious how the implementation code, then:

Scanner v2/StackRox Scanner: https://github.com/stackrox/scanner
Scanner V4: https://github.com/stackrox/stackrox/tree/master/scanner

If you just care about the Helm charts, then we make the distrinction with `-v4-` in the file names. In the https://github.com/stackrox/stackrox/tree/master/image/templates/helm/shared/templates directory, you'll find the Scanner V4 files ranging from 02-scanner-v4-00-serviceaccount.yaml to 99-scanner-v4-openshift-monitoring.yaml. Similarly, you'll find Scanner V4-related files https://github.com/stackrox/stackrox/tree/master/image/templates/helm/shared/config-templates

acravn commented:
Hi @mclasmeier  sorry for the delay on updating this PR. Hopefully it looks better after my latest push. 

I did the following:

- Added unit tests for both central and secured cluster services
- Fixed the formatting with the if statements starting with {{- }}
- Re-did my approach for pre-existing TLS secrets. It now will changes the volume configuration to mount the pre-existing secrets if provided, I did need to still use a lookup to get the ca.pem from the existing ca secret since it's required for the validating webhook client config. I was trying to use the "srox.safeLookup" but I couldn't seem to get it to work as expected.

Thanks!

acravn commented:
/retest-required

RTann commented:
@acravn can you update the Scanner V4 configuration files as well? Scanner v2 will be deprecated in favor of Scanner V4 in the future, and we'd prefer any future development towards StackRox vulnerability scanners to be done for Scanner V4, instead

porridge commented:
@acravn what is the status of this PR? Do you still intend to work on it?
